typing/typecore.ml

```diff
@@ -4738,6 +4749,31 @@ and type_cases
       )
       half_typed_cases
   in
+  let cases =
+    let may_backtrack = does_contain_gadt && not !Clflags.principal in
+    if not may_backtrack then mk_cases false else
+    let snap = Btype.snapshot () in
+    let has_equation_escape err =
+      match trace_of_error err with
+        Some tr ->
+          List.exists Ctype.Unification_trace.
```

Comment:
Nitpick: I find it fairly somewhat confusing to have a dot followed by whitespace. What about the following?

```ocaml
let is_escape =
  let open Ctype.Unification_trace in
  function
  | Escape {kind = Equation _} -> true
  | _ -> false
in
List.exists is_escape tr
```